### PR TITLE
New implementation of Tabby

### PR DESCRIPTION
**Pull request by RamonGilabert**
All right, here it is. The new implementation of Tabby. It features:
- Behaviors (a way to add custom behaviors to the bar).
- A Collection View to easy add stuff into the it.
- Different constants and properties.
- Different set of animations.

With all that, there's one missing thing that you'll see in terms of a `TODO`. That is, find the difference when you add new items and do whatever is best in the tab bar. That is, delete, insert or reload all the way. Something similar to what we have in Spots.

I believe, though, that we could just put it right away. And then add it later.

It'd also be nice to set up Carthage.

app

What do you think, @hyperoslo/ios?

### CODE REVIEW

File: Sources/Cells/TabbyCell.swift [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    setupConstraints()
+  }
+
+  // MARK: - Helper methods
+
+  func handleBehaviors(selected: Bool) {
+    switch Constant.Behavior.labelVisibility {
+    case .Invisible:
+      label.alpha = 0
+      break
```
Comment (zenangst) [inline Sources/Cells/TabbyCell.swift]: You could drop the `break`

File: Sources/Cells/TabbyCell.swift [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    case .ActiveVisible:
+      label.alpha = selected ? 1 : 0
+      break
```
Comment (zenangst) [inline Sources/Cells/TabbyCell.swift]: Same here

File: Sources/TabbyController.swift [hunk 3]
```diff
@@ -17,24 +17,28 @@ public class TabbyController: UIViewController {
...
-   - Parameter controller: The actual controller, can be any.
-   - Parameter kind: The image that will appear in the tab bar.
+   - Parameter controller: The controller that you set as the one that will appear when tapping the view.
+   - Parameter image: The image that will appear in the TabbyBarItem.
    */
-  public var controllers: [(controller: UIViewController, image: UIImage?)] = [] {
+  public var items: [TabbyBarItem] {
     didSet {
-      tabbyBar.prepare(controllers)
+      var currentItem = tabbyBar.items[index]
```
Comment (zenangst) [inline Sources/TabbyController.swift:37]: Should we care about checking if it is not out of bounds here?

Comment (RamonGilabert) [reply Sources/TabbyController.swift:37]: We totally should.

File: Sources/Views/TabbyBar.swift [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+   Initializer
+   */
+  public init(items: [TabbyBarItem]) {
+    self.items = items
+    
+    super.init(frame: .zero)
+
+    backgroundColor = Constant.Color.background
+
+
```
Comment (zenangst) [inline Sources/Views/TabbyBar.swift]: You can remove this newline

File: Sources/Views/TabbyBar.swift [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+  var delegate: TabbyBarDelegate?
+
+  // MARK: - Initializers
+
+  /**
+   Initializer
+   */
+  public init(items: [TabbyBarItem]) {
+    self.items = items
+
```
Comment (zenangst) [inline Sources/Views/TabbyBar.swift]: And this newline

File: Sources/Views/TabbyBar.swift [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+ The actual tab bar.
+ */
+public class TabbyBar: UIView {
+
+  static let collectionObserver = "contentSize"
+  static let KVOContext = UnsafeMutablePointer<()>(nil)
+
+  lazy var translucentView: UIVisualEffectView = UIVisualEffectView(effect: UIBlurEffect(style: .Light))
+
+  lazy var layout: TabbyLayout = { [unowned self] in
```
Comment (zenangst) [inline Sources/Views/TabbyBar.swift]: You are not referencing `self` here so you can just drop the `[unowned self]`

Comment (RamonGilabert) [reply Sources/Views/TabbyBar.swift]: I totally was before.

File: Sources/Cells/TabbyCell.swift [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+      label.alpha = selected ? 1 : 0
+    default:
+      label.alpha = 1
+    }
+  }
+
+  // MARK: - Constraints
+
+  func setupConstraints() {
+    let offset: CGFloat = label.alpha == 1 ? 8 : 0
```
Comment (zenangst) [inline Sources/Cells/TabbyCell.swift:57]: What is `8` here?

Comment (RamonGilabert) [reply Sources/Cells/TabbyCell.swift:57]: A beautiful magic number... Maybe it should be a configurable constant. Thought about it.

File: Sources/Items/TabbyBarItem.swift [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+  public init(controller: UIViewController, image: UIImage?, animation: TabbyAnimation.Kind = Constant.Animation.initial) {
+    self.controller = controller
+    self.image = image
+    self.animation = animation
+  }
+}
+
+public func ==(lhs: TabbyBarItem, rhs: TabbyBarItem) -> Bool {
+  return lhs.controller == rhs.controller
+    && lhs.image == rhs.image
```
Comment (zenangst) [inline Sources/Items/TabbyBarItem.swift:18]: Can you compare images like this? 🤔

Comment (zenangst) [inline Sources/Items/TabbyBarItem.swift:18]: Same goes for the controller I guess

File: Sources/Views/TabbyBar.swift [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+  }
+
+  var selectedItem: Int = 0 {
+    didSet {
+      positionIndicator(selectedItem)
+      collectionView.reloadData()
+    }
+  }
+
+  var delegate: TabbyBarDelegate?
```
Comment (zenangst) [inline Sources/Views/TabbyBar.swift]: Delegates should we `weak` right?

File: Sources/Views/TabbyBar.swift [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+  }
+
+  // MARK: - Collection View setup
+
+  func setupCollectionView() {
+    collectionView.delegate = self
+    collectionView.dataSource = self
+    collectionView.backgroundColor = Constant.Color.collection
+
+    collectionView.registerClass(
```
Comment (zenangst) [inline Sources/Views/TabbyBar.swift:106]: Why the newline here?

File: Sources/TabbyController.swift [hunk 11]
```diff
@@ -96,25 +116,25 @@ public class TabbyController: UIViewController {
    */
   public var delegate: TabbyDelegate?
```
Comment (zenangst) [inline Sources/TabbyController.swift]: Same for this `delegate`, should also be annotated as `weak` right?

File: Sources/Items/TabbyBarItem.swift (hunk 8 above)
Comment (RamonGilabert) [reply Sources/Items/TabbyBarItem.swift:18]: It sorta works. Don't know. But it's basically the same reference.

File: Sources/Views/TabbyBar.swift (hunk 10 above)
Comment (RamonGilabert) [reply Sources/Views/TabbyBar.swift:106]: I like to separate property setting and method calling.

File: Sources/Cells/TabbyCell.swift (hunk 7 above)
Comment (RamonGilabert) [reply Sources/Cells/TabbyCell.swift:57]: What do you think?

**Comment by zenangst** [comment]
Should we drop the prefix in the naming?

**Comment by RamonGilabert** [comment]
I don't want to be calling TabbyController -> Controller or TabbyBar -> Bar.

**Comment by zenangst** [comment]
Good point, screw it. I like this PR... MERGE!

**Comment by zenangst** [comment]
giphy-79

**Comment by RamonGilabert** [comment]
truth-bombs-rick-morty-16-gifs-6

**Pull request merged**